blockchain/v2/processor.go

```diff
@@ -0,0 +1,191 @@
+	"github.com/tendermint/tendermint/types"
+)
+
+type queueItem struct {
+	block  *types.Block
+	peerID p2p.ID
+}
+
+type blockQueue map[int64]queueItem
+
+type pcState struct {
+	height       int64
+	queue        blockQueue
+	chainID      string
+	blocksSynced int
+	draining     bool
+	tdState      tdState.State
+	context      processorContext
+}
+
+func (ps *pcState) String() string {
+	return fmt.Sprintf("queue: %d draining: %v, \n", len(ps.queue), ps.draining)
+}
+
+// newPcState returns a pcState initialized with the last verified block enqueued
+func newPcState(initBlock *types.Block, tdState tdState.State, chainID string, context processorContext) *pcState {
+	var myID p2p.ID = "thispeersid" // XXX: this should be the nodes peerID
+	initItem := queueItem{block: initBlock, peerID: myID}
+
+	return &pcState{
+		height:       initBlock.Height,
+		queue:        blockQueue{initBlock.Height: initItem},
+		chainID:      chainID,
+		draining:     false,
+		blocksSynced: 0,
+		context:      context,
+		tdState:      tdState,
+	}
+}
+
+// nextTwo returns the next two unverified blocks
+func (state *pcState) nextTwo() (queueItem, queueItem, error) {
+	if first, ok := state.queue[state.height+1]; ok {
+		if second, ok := state.queue[state.height+2]; ok {
+			return first, second, nil
+		}
+	}
+	return queueItem{}, queueItem{}, fmt.Errorf("not found")
+}
+
+// synced returns true when only the last verified block remains in the queue
+func (state *pcState) synced() bool {
+	return len(state.queue) == 1
+}
+
+func (state *pcState) advance() {
+	delete(state.queue, state.height)
+	state.height++
+	state.blocksSynced++
+}
+
+func (state *pcState) enqueue(peerID p2p.ID, block *types.Block, height int64) error {
+	if _, ok := state.queue[height]; ok {
+		return fmt.Errorf("duplicate queue item")
+	}
+	state.queue[height] = queueItem{block: block, peerID: peerID}
+	return nil
+}
+
+// purgePeer moves all unprocessed blocks from the queue
+func (state *pcState) purgePeer(peerID p2p.ID) {
+	// what if height is less than state.height?
+	for height, item := range state.queue {
+		if item.peerID == peerID {
+			delete(state.queue, height)
+		}
+	}
+}
+
+type peerError struct {
+	priorityHigh
+	peerID p2p.ID
+}
+
+type pcDuplicateBlock struct {
+	priorityNormal
+}
+
+type bcBlockResponse struct {
+	priorityNormal
+	peerID p2p.ID
+	block  *types.Block
+	height int64
+}
+
+type pcBlockVerificationFailure struct {
+	priorityNormal
+}
+
+type pcBlockProcessed struct {
+	priorityNormal
+	height int64
+	peerID p2p.ID
+}
+
+type pcProcessBlock struct {
+	priorityNormal
+}
+
+type pcStop struct {
+	priorityNormal
+}
+
+type pcFinished struct {
+	priorityNormal
+	height       int64
+	blocksSynced int
+	state        tdState.State
+}
+
+func (p pcFinished) Error() string {
+	return "finished"
+}
+
+// TODO: timeouts
+// XXX: we are mutating state here, is that what we want?
+func pcHandle(event Event, state *pcState) (Event, *pcState, error) {
```

Comment:
Does it make sense to rather have this as a method on `pcState`? e.g.

```go
func (state *pcState) handle(event Event) (Event, error) {
    // ...
}
```

This suggested approach will allow us to extract the code under each `case` statement under the event `switch` into its own function, e.g. `handleBlockResponse`, `handleProcessBlock`, etc. I could make these kinds of functions easier to read.